hwdef: Add MambaF405 target

jeanphilippehell opened the pull request:
Hi this is my first PR :)

I've added the hwdef to support the MambaF405 board, as discussed here :
[https://github.com/ArduPilot/ardupilot/issues/14301](https://github.com/ArduPilot/ardupilot/issues/14301)

I have tested the following  :

  -  IMU
  -  Sbus input
  -  RC out 1-4
  -  Battery voltage
  -  Dataflash

The board has flown with Plane 4.0.5 then to make this PR I've rebased my work on the current master.
In the meantime the init without baro has been [fixed](https://github.com/ArduPilot/ardupilot/pull/14317) in master so I removed my redundant commits about that.

I've build and tested for this specific target on master :
- Bootloader
- Plane
- Copter

Note that the board ID 1002 is shared with the omnibusf4 target, and the plane / copter binary works indifferently with the omnibusf4 or the MambaF405 bootloader.
I don't know if it is necessary to assign a new board ID for the MambaF405 as there is no added value in my hwdef-bl for now.

Thanks

JP

andyp1per commented:
Yeah, you need to request a new id

jeanphilippehell commented:
Ok thanks, where do I need to request id ? Gitter ?

Sekilsgs2 commented:
You dont fix i2c and uart 3

andyp1per commented:
@jeanphilippehell  Add a change to https://github.com/ardupilot/ardupilot/blob/master/Tools/AP_Bootloader/board_types.txt to get the new ID

auturgy commented:
MambaF405 already has an ID (1019).

Regards,

James 

> On 22 Sep 2020, at 5:14 pm, Andy Piper <notifications@github.com> wrote:
> 
> ﻿
> @jeanphilippehell Add a change to https://github.com/ardupilot/ardupilot/blob/master/Tools/AP_Bootloader/board_types.txt to get the new ID
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

andyp1per commented:
Ah my mistake, thanks James!

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef-bl.dat:
```diff
@@ -0,0 +1,39 @@
+MCU STM32F4xx STM32F405xx
+
+# board ID for firmware load
+APJ_BOARD_ID 1002
```
Change to 1019

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,141 @@
+HAL_CHIBIOS_ARCH_F405 1
+
+# board ID for firmware load
+APJ_BOARD_ID 1002
```
Change to 1019

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,141 @@
+I2C_ORDER I2C2
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART6 USART1
```
Only two UARTS? surely there are more than this?

andyp1per commented:
@jeanphilippehell you will need to build the bootloader as part of this change - look in Tools/bootloaders

jeanphilippehell commented:
Hi, thanks for the feedback

I've fixed my commit after doing the following : 
- fixed I2C definitions and added UART3 (@Sekilsgs2 would you like to test this since I have no I2C or UART device soldered)
- changed the board Id to 1019 in the bootloader and ardupilot
- compiled and flashed the BL and Arduplane to verify the upload

JP

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,145 @@
+I2C_ORDER I2C1
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART6 USART3 USART1
```
This order dictates what SERIAL1/2/3 will be - have a think about whether its better to align the numbers with the documented UART numbers. You can put gaps in this list.

jeanphilippehell replied:
If we start from OTG (USB) and go clockwise, the order of the serial ports on the board is indeed OTG1 USART6 USART3 USART1
![See https://cdn.shopify.com/s/files/1/0027/2708/4144/files/19_F405.jpg](https://cdn.shopify.com/s/files/1/0027/2708/4144/files/19_F405.jpg)

jeanphilippehell replied:
Ok I get it, we may want to align the number of the UARTS on the board with the SERIAL number on ArduPilot
I check if there is a convention or something

andyp1per replied:
Right so currently someone will see TX6/RX6 and try and set SERIAL6 and get confused (maybe)

jeanphilippehell replied:
All right so from https://ardupilot.org/plane/docs/parameters.html#serial-parameters I get the SERIAL order, along with how I intend to link them :
SERIAL0 => USB      => OTG1
SERIAL1 => Telem1  =>  USART6 (Still wondering if to do so knowing that on this board USART6 is tied to the ESCs connector, and usually used for ESC telemetry)
SERIAL2 => Telem2  => not present on this board
SERIAL3 => GPS       => USART3
From SERIAL4 there is no specific assignment on Ardupilot, if I am right

jeanphilippehell replied:
and USART1 is used for the RC in so I don't know which SERIAL to use for this one

jeanphilippehell replied:
Ok I will push this :
SERIAL0 => OTG1 (USB)
SERIAL1 => USART1 (SBUS / PPM in / TX/RX1)
SERIAL2 => EMPTY
SERIAL3 => USART3 (TX/RX3)
SERIAL4 => EMPTY
SERIAL5 => EMPTY
SERIAL6 => USART6 (ESC Telem / TX/RX6)

andyp1per replied:
Makes sense to me - I suspect anything else is going to be confusing

andyp1per commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,145 @@
+PC5 VBUS INPUT OPENDRAIN
+
+# USART1 is wired to SBUS input
+PA10 TIM1_CH3 TIM1 RCININT FLOAT LOW
```
Do you want to provide an alt config so that you can use this as a regular uart? Look at the KakauteF7Mini hwdef fro an example

jeanphilippehell replied:
Nice feat, I'll do that !

vierfuffzig commented:
@jeanphilippehell @andyp1per add I2C2 on PB10 / PB11 as alt config?

jeanphilippehell commented:
> @jeanphilippehell @andyp1per add I2C2 on PB10 / PB11 as alt config?

Checking right now, I didn't see your message

Sekilsgs2 commented:
Mamba have inverter - im think need add info about which pin and comment or add with alt..

jeanphilippehell commented:
> Mamba have inverter - im think need add info about which pin and comment or add with alt..

Right, I just catched it : PC0
The idea is to have an alt setting with UART1 inverted and another alt with UART1 non inverted ?

Sekilsgs2 commented:
> > Mamba have inverter - im think need add info about which pin and comment or add with alt..
> 
> Right, I just catched it : PC0
> The idea is to have an alt setting with UART1 inverted and another alt with UART1 non inverted ?

Yes, but im think this not actual - because rc in only on uart1 and not other -  maybe we can make ppm pin for other rc in variants with other alt - and with ppm in active for rc in we can use uart1 as normal serial port for many things...

jeanphilippehell commented:
> > > Mamba have inverter - im think need add info about which pin and comment or add with alt..
> > 
> > 
> > Right, I just catched it : PC0
> > The idea is to have an alt setting with UART1 inverted and another alt with UART1 non inverted ?
> 
> Yes, but im think this not actual - because rc in only on uart1 and not other - maybe we can make ppm pin for other rc in variants with other alt - and with ppm in active for rc in we can use uart1 as normal serial port for many things...

True.
For now I have defined the following ALT settings :
ALT 0 (default) : RX1 is RCININT, RX/TX6 are USART6
ALT 1: RX1 is RX1, RX/TX6 are USART6
ALT 2: RX1 is RCININT, RX/TX6 are I2C2
ALT 3: RX1 is RX1, RX/TX6 are I2C2

I could add SBUS_INVERT_RX as a separate ALT, knowing that we have eleven possibilities (0 to 10)

jeanphilippehell commented:
What I did :
- Rebase to the last master
- Change serial order to be consistent
- Add ALT settings as described in previous comment and code comment

I tested :
- Plane, on MambaF405 with Frsky SBUS soldered on SBUS input
- Build, flash, boot, Mavlink

Also tested all ALT settings: 
- When RX1 is RCININT the inversion does not affect receiving the signals. Maybe it is proper of the SBUS protocol
- When RX1 is RX1, and SERIAL1 is set to RCININT via mavlink, the RC only works when RX1 is not inverted (ALT 2 and 6). This means to me that the inversion _does_ work

Didn't tested the I2C ports

andyp1per reviewed (approved):
LGTM. You will need to build the BL firmware to go in the PR.

jeanphilippehell commented:
> LGTM. You will need to build the BL firmware to go in the PR.

@andyp1per I have three ouputs when I build the bootloader :
- build/MambaF405/bin/AP_Bootloader.apj
- build/MambaF405/bin/AP_Bootloader.bin (the one I flashed on my board with dfu-utils) : do I copy it to Tools/bootloaders/MambaF405_bl.bin ?
- build/MambaF405/bootloader/AP_Bootloader : is this the .hex ? do I copy it to Tools/bootloaders/MambaF405_bl.hex ?
Thanks

andyp1per commented:
@jeanphilippehell just copy the .bin file. If you then rebuild you should get the correct boot loader embedded in your firmware and can then try upgrading via MP or mavproxy

jeanphilippehell commented:
> @jeanphilippehell just copy the .bin file. If you then rebuild you should get the correct boot loader embedded in your firmware and can then try upgrading via MP or mavproxy

@andyp1per sorry I don't understand what is requested for the bootloader. 
To build the bootloader I use the command :
`./waf configure --board=MambaF405 --bootloader && ./waf bootloader`
I've then copied the AP_Bootloader.bin generated as Tools/bootloaders/MambaF405_bl.bin, and added it in my commit

andyp1per commented:
I think you can just do ```./Tools/scripts/build_bootloaders.py MambaF405``` that will build the boot loaders and boot them in the right place, you can then just git add

jeanphilippehell commented:
Ok I've generated all the bootloader files with `./Tools/scripts/build_bootloaders.py MambaF405` and added them to the commit
The step missing was the intel hex generation with bin2hex.py with an offset of 0x08000000
Anyway thanks for pointing out and the help @andyp1per

IamPete1 commented:
I guess you will also want to add it here so we get builds on the firmware server, https://github.com/ArduPilot/ardupilot/blob/9bde294112bb0726134c6b012b5027526670f469/Tools/scripts/build_binaries.py#L620

andyp1per commented:
@IamPete1 @jeanphilippehell yes and also in ```Tools/scripts/generate_manifest.py```

jeanphilippehell commented:
@IamPete1 @andyp1per Thanks, I've added the MamabaF405 in [Tools/scripts/build_binaries.py](https://github.com/ArduPilot/ardupilot/pull/15392/commits/90fbc0ffe09c69318aec01b94637efacec13d279#diff-dcc8fd382ee9b25a9fb237e78b9d9526) and [Tools/scripts/generate_manifest.py](https://github.com/ArduPilot/ardupilot/pull/15392/commits/90fbc0ffe09c69318aec01b94637efacec13d279#diff-b9793f4f1929f363b1738555dbf89165)

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,168 @@
+# one IMU rotated in yaw
+IMU Invensense SPI:mpu6000 ROTATION_YAW_180
+
+define CONFIG_HAL_BOARD_SUBTYPE HAL_BOARD_SUBTYPE_CHIBIOS_GENERIC
```
don't add SUBTYPE, just remove this line and it will be good

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,168 @@
+I2C_ORDER I2C1 I2C2
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART1 EMPTY USART3 EMPTY EMPTY USART6
```
why is USART6 right at the end? why not in as SERIAL2 ?

andyp1per replied:
As discussed its sufficient to simply document the current order in the hwdef.dat

andyp1per replied:
Perhaps add a comment that this is the order in the documentation

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,168 @@
+
+# crystal frequency
+OSCILLATOR_HZ 8000000
+STM32_PLLM_VALUE 8
```
you shouldn't need the PLLM value

tridge commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef-bl.dat:
```diff
@@ -0,0 +1,40 @@
+
+# crystal frequency
+OSCILLATOR_HZ 8000000
+STM32_PLLM_VALUE 8
```
remove PLLM value

tridge reviewed (changes requested):
a README.md in the folder for the board explaining the pinouts and the alt config settings would be nice, see example in CubeBlack

andyp1per commented:
Hi @jeanphilippehell - if you can just make the changes tridge has asked for we can get this merged!

jeanphilippehell commented:
Hi, sorry for the delay ...

I've force-pushed the modifications :
- Added a README.md (might benefit from review / maturity and user feedback as I am not able to test all the periphs and I seem to own an older version of the Mamba F405 than the one in the picture)
- Made the modification requested by tridge
- Added a comment explaining the order of SERIAL ports vs labels on the board
- Rebased on the last master
- Build Bootloader and Plane, flashed both, checked runtime
- Diff'ed Bootloader with the one previously commited, no differences

_But_

I ran into trouble configuring the board with `./waf configure --board MambaF405`
There seems to be an issue with this [commit](https://github.com/ArduPilot/ardupilot/commit/46976c43583d0d9afef352c809472b84050b9439)
In fact I seem to be the only one using ALT settings for an OUTPUT type gpio
The parsing in **chibios_hwdef.py** line 547 does not work when there is no number in the type of GPIO (in my case "OUTPUT")
In this file, in the test line 546
`        if result:`
 `           return int(result.group(1))`
the variable `result` does exists, but does not contains any value. Thus the call `result.group(1)` returns an empty field which is not accepted in the cast to int.
Since I don't know exactly how the alternate gpio configuration works at this level, I have no fix to offer.
My way to build the binaries and test them was to declare OUTPUT1 instead of OUTPUT in the hwdef.dat
But since this does not seem like an acceptable fix, I've put back the OUTPUT declaration like on the previous commits before this last rebase.

JP

tridge commented:
@jeanphilippehell I don't think you need all those alt configs. What you're trying to achieve is better done with lines like this one:

https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_HAL_ChibiOS/hwdef/Pixracer/hwdef.dat#L147

in that one a GPIO pin is marked as controlling inversion of a uart pin. You can mark any GPIO as being an inversion control pin for a RX or TX pin by using the special name UARTn_TXINV and UARTn_RXINV.
When you do that then the SERIALn_OPTIONS bits for inversion will work just like they do for the F7/H7 MCUs.

andyp1per commented:
@jeanphilippehell are you able to make these changes?

GHswitt commented on libraries/AP_HAL_ChibiOS/hwdef/MambaF405/hwdef.dat:
```diff
@@ -0,0 +1,168 @@
+# reserve 32k for bootloader and 32k for flash storage
+FLASH_RESERVE_START_KB 64
+
+# define default battery setup
+define HAL_BATT_VOLT_PIN 12
+define HAL_BATT_CURR_PIN 11
```
Should that not be 11 and 13 (see #adc above)?

andyp1per commented:
@jeanphilippehell bump - please let us know if you can't make these changes. I would do them but have no hardware to test on.

Sekilsgs2 commented:
> @jeanphilippehell bump - please let us know if you can't make these changes. I would do them but have no hardware to test on.

Yes yes - very slow man :)
Make changes - im have this board and can test.

GHswitt commented:
I've the hardware and I'm also currently testing it, with changes.

andyp1per commented:
@GHswitt so if you can clone this branch and create a new PR with this and your changes + the changes tridge wants that would be awesome. But don't squash @jeanphilippehell 's changes into yours so that he gets credit in the commit history.

GHswitt commented:
I can do that, but first I want to check that everything is working. For example the battery current monitoring doesn't work and I want to test another UART for telemetry. And most important I didn't test it in flight yet...

GHswitt commented:
It seems I have the F405 Mk2, which is a bit different. For example I need AHRS_ORIENTATION of Yaw270, even if the arrow of the board shows to the front. Also the ADC pins are a little bit different. 

How should I proceed?

andyp1per commented:
Clone, PR and rename MambaF405v2 - if this PR ever goes in it can be v1

peterbarker commented:
I might close this one, I think.

If support for the v1 board was wanted it would probably be done by inheritting from the v2 (or vice-versa), or perhaps a common base.

Thanks for adding the new board, 'though!
